### PR TITLE
NIFI-7740: Add Records Per Transaction and Transactions Per Batch properties to PutHive3Streaming

### PR DESCRIPTION
**Pull request by mattyb149**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Adds the above properties to PutHive3Streaming (PH3S) to be feature-compliant and user-configurable for transaction properties for PH3S that were available for PutHiveStreaming. These tuning properties can help avoid heartbeat/timeout issues with very large files going through PH3S, the idea is that you reduce the number of Records Per Transaction and increase the Transactions Per Batch to ensure that all records get committed to Hive successfully.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [x] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/hive/streaming/RecordsEOFException.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hive.streaming;
+
+/**
+ * This is a "marker class" used by the HiveRecordWriter to indicate there are no more records in the input stream.
+ * It is used by PutHive3Streaming to determine that all records have been written to transaction(s). IT extencs
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/hive/streaming/RecordsEOFException.java:21]: Typo: "IT extencs"

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/nifi/processors/hive/PutHive3Streaming.java [hunk 2]
```diff
@@ -171,6 +172,28 @@
             .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
             .build();
 
+    static final PropertyDescriptor RECORDS_PER_TXN = new PropertyDescriptor.Builder()
+            .name("hive-stream-records-per-transaction")
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/nifi/processors/hive/PutHive3Streaming.java:176]: All others properties have "hive3-" prefix here. I think these two should follow the same pattern.

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/nifi/processors/hive/PutHive3Streaming.java [hunk 3]
```diff
@@ -171,6 +172,28 @@
...
+            .required(true)
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+            .addValidator(StandardValidators.NON_NEGATIVE_INTEGER_VALIDATOR)
+            .defaultValue("0")
+            .build();
+
+    static final PropertyDescriptor TXNS_PER_BATCH = new PropertyDescriptor.Builder()
+            .name("hive-stream-transactions-per-batch")
+            .displayName("Transactions per Batch")
+            .description("A hint to Hive Streaming indicating how many transactions the processor task will need. The product of Records Per Transaction (if not zero) "
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/nifi/processors/hive/PutHive3Streaming.java:188]: Minor typo: Records per Transaction (in order to use the same case as in the display name of the referenced property)

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/hive/streaming/RecordsEOFException.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ */
+public class RecordsEOFException extends SerializationError {
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/hive/streaming/RecordsEOFException.java:23]: Hive's exception hierarchy looks a bit strange to me (`SerializationError` which is an `Exception` in fact) and we need to pass a dummy `new Exception()` in the constructor.
Would not it be better to descend from `RuntimeException` instead?

**Review by turcsanyip — COMMENTED** [review]
@mattyb149 Found some typos and minor issues so far.
Will test it with a Hive server later.

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/hive/streaming/RecordsEOFException.java (hunk 1 above)
Comment (mattyb149) [reply nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/hive/streaming/RecordsEOFException.java:21]: Whoa, must've been typing in the wrong window :) Will remove

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/nifi/processors/hive/PutHive3Streaming.java (hunk 2 above)
Comment (mattyb149) [reply nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/nifi/processors/hive/PutHive3Streaming.java:176]: Good catch, will update

File: nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/hive/streaming/RecordsEOFException.java (hunk 4 above)
Comment (mattyb149) [reply nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/hive/streaming/RecordsEOFException.java:23]: Probably, I'm not even sure why I picked that (but I did have to choose the same package name in order to use the constructor I think). Will change to RuntimeException

Comment (turcsanyip) [reply nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/hive/streaming/RecordsEOFException.java]: My idea was wrong, `RuntimeException`-s are handled differently by the Hive client and the transaction would be marked dead:
https://github.com/apache/hive/blob/bfb55568bcaa8e91112f4e5b983edcc9dcd2844c/streaming/src/java/org/apache/hive/streaming/HiveStreamingConnection.java#L814
Please put back the original `SerializationError`.

Comment (mattyb149) [reply nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/hive/streaming/RecordsEOFException.java]: That must've been why I picked that :) Will restore

**Review by turcsanyip — APPROVED** [review]
Tested with a simple flow connecting to Hive and loading data via Hive Streaming. Works as expected.

+1 LGTM
Merging to main.

**Pull request closed without merging**